Implementação do requisito 2

mathpablo opened the pull request.

mathpablo committed:
Criação do enum para tipos de exames.

lucasferrarisoares committed:
Criando classe médico

mathpablo committed:
Criação da classe exame, e dos enus exames e tipos exames

mathpablo committed:
Merge branch 'Req-1' into Req-2

lucasferrarisoares committed:
Update DoctorModel.java

lucasferrarisoares committed:
Finalizado requisição

Parte base da requisição finalizada, agora entro na fase de revisão/melhoria do código

mathpablo committed:
Adicionei a pasta examModel, e fiz algumas alterações na pasta patientModel

lucasferrarisoares committed:
Update AdmissionLogService.java

mathpablo committed:
Merge branch 'Req-1' of https://github.com/CarlosFavarao/ControleDePacientes into Req-2

lucasferrarisoares committed:
Criado Histórico de médicos por internação.

mathpablo committed:
adicionei o método de deletar exames na ExamController, na PatientModel tive que adicionar a coluna admission na tabela pacientes, na classe AdmissionLogService tive que colocar uma validação para quando der alta para o paciente atualizar a tabela de paciente assim deixando livre para outra internação

mathpablo committed:
adicionei para buscar todos os exames que aquele paciente tem pelo id

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+    private PatientRepository patientRepository;
+
+    @Autowired
+    private DoctorRepository doctorRepository;
```
Nesse caso, chamar as Services em vez de Repository, para que se use os métodos já refinados.

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+                .orElseThrow(() -> new RuntimeException("Paciente não encontrado."));
+
+        DoctorModel doctorModel = doctorRepository.findById(dto.getDoctorId())
+                .orElseThrow(() -> new RuntimeException("Doutor não encontrado."));
```
Fazer o tratamento de Exeptions nas Services referentes às classes usadas aqui, no caso PatientService e DoctorService

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+        boolean conflito = examRepository.existsByPatientAndDateTime(patientModel, dto.getDateTime());
+        if (conflito) {
+            throw new RuntimeException("Já existe um exame agendao nesse horario para esse paciente");
+        }
```
if(examRepository.existsByPatientAndDateTime(patientModel, dto.getDateTime()){
    throw new RuntimeException("Já existe um exame agendao nesse horario para esse paciente");
}

Essa maneira elimina a necessidade da variável conflito

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+    @Transactional
+    public void deleteExam(Long id) {
+        ExamModel exam = examRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Exame não encontrado com id " + id));
```
Criar um método FindById nessa service e, nele, realizar o tratamento de Exceção. Depois fazer apenas this.findById(id), seguindo com a exclusão normalmente à partir daí.

CarlosFavarao commented on src/main/resources/application.properties:
Fazer o método FindById, pois seu uso é indispensável e terminar o CRUD com um método para edição de exames (update) (ExamService).

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+        if (!patientModel.getResponsibleDoctor().getId().equals(doctorModel.getId())) {
+            throw new RuntimeException("Médico solicitante não é o responsável pelo paciente.");
+        }
+        Long doctorId = patientModel.getResponsibleDoctor().getId();
```
Não usado

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/model/ExamModel.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ManyToOne
+    @JoinColumn(name = "doctor_id", nullable = false)
+    private DoctorModel doctor;
```
Adicionar o (fetch = FetchType.LAZY) na anotação @ManyToOne, para prevenir buscas excessivas no Banco e retornos com Loop infinito

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/dto/ExamDto.java:
Adicionar o nome do paciente e do médico, para que o retorno seja mais completo

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/controller/ExamController.java:
Adicionar um @PutMapping para possibilitar a Edição de Exames (Update) e um @GetMapping com o FindById.

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/controller/ExamController.java:
```diff
@@ -0,0 +1,45 @@
+        @Autowired
+        private ExamService examService;
+
+        //Criação de exame
+        @PostMapping("/schedule")
+        public ResponseEntity<String> schedule(@RequestBody ExamDto dto) {
+            try {
+                examService.ScheduleExam(dto);
+                return ResponseEntity.status(HttpStatus.CREATED).body("Exame agendado com sucesso.");
+            } catch (RuntimeException e) {
+                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
+            }
+        }
```
Alinhar o código dessa área com a parte mais abaixo, ela está com espaços a mais.

lucasferrarisoares replied:
Colocar regra de negócios na service ao invés do controller

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/controller/ExamController.java:
```diff
@@ -0,0 +1,45 @@
+            }
+        }
+
+        //Deletar exame
```
Excluir comentários desnecessários, nesse caso o próprio nome do método já é autoexplicativo

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+    @Transactional
+    public List<ExamDto> listExamsByPatients(Long patientId) {
+        PatientModel patient = patientRepository.findById(patientId)
+                .orElseThrow(() -> new RuntimeException("Paciente não encontrado."));
```
Utilizar o próprio findById que será criado posteriormente nessa classe, realizando o tratamento de Exceção nele.

CarlosFavarao commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+            dto.setPatientId(exams.getPatient().getId());
+            dto.setDoctorId(exams.getDoctor().getId());
+            return dto;
+        }).collect(Collectors.toList());
```
Aqui você pode criar um método na classe ExamDto e utilizá-lo sempre que precisar converter esses dados, diminuindo o tamanho do código e mantendo as responsabilidades de cada classe.

CarlosFavarao reviewed (changes requested):
Alterações para melhorar a qualidade do código

lucasferrarisoares committed:
adicionado nome a classe Doctor

lucasferrarisoares commented on src/main/java/br/com/ControleDePacientes/model/PatientModel.java:
```diff
@@ -17,4 +19,10 @@ public class PatientModel {
 
     @Column(nullable = false)
     private String name;
+
+    @Column
+    private LocalDateTime Admission;
+
+    @ManyToOne
+    private DoctorModel responsibleDoctor;
```
Esses dados são necessários? Visto que as duas informações constam no log de internação do paciente

lucasferrarisoares commented on src/main/java/br/com/ControleDePacientes/controller/ExamController.java:
```diff
@@ -0,0 +1,45 @@
+    //Listar exames por paciente
+    @GetMapping("/patient/{patientId}")
+    public ResponseEntity<List<ExamDto>> listByPatients(@PathVariable Long patientId){
+           List<ExamDto> exam = examService.listExamsByPatients(patientId);
+           return ResponseEntity.ok(exam);
```
o return poderia ser simplificado da seguinte forma: "return ResponseEntity.ok(examService.listExamsByPatients(patientId));"
não sendo necessários a definição do List exam

lucasferrarisoares committed:
Mudanças pós reunião

Mudança no nome do atributo que contem um DoctorModel dentro de AdmissionLogModel
Nome do médico adicionado em DoctorModel
Colocado ResponseEntity no return do controller de DoctorController
Optional no retorno de DoctorRepository
Mudança de DoctorHistory... para DoctorLog para uma padronização do projeto

EwertonYamada commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+
+        if (patientModel.getAdmission() == null) {
+            throw new RuntimeException("Paciente não está internado. ");
+        }
```
Conforme o Lucas comentou, talvez não seria necessário ter a data de internação no paciente. Essa validação pode ser feita verificando se no admissionlog existe alguma internação com esse paciente e que a data de alta é null.

EwertonYamada commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+
+        if (!patientModel.getResponsibleDoctor().getId().equals(doctorModel.getId())) {
+            throw new RuntimeException("Médico solicitante não é o responsável pelo paciente.");
+        }
```
Conforme o Lucas comentou, talvez não seria necessário ter o doctor no patient. Daria para validar tanto esse ponto quanto o anterior em uma unica verificação: criar um select count que retorna true ou false onde no log de internação o id do paciente é igual ao que veio no dto e o id do médico é igual ao que veio no dto e onde a data de alta é null

EwertonYamada commented on src/main/java/br/com/ControleDePacientes/service/ExamService.java:
```diff
@@ -0,0 +1,100 @@
+        examModel.setPatient(patientModel);
+        examModel.setDoctor(doctorModel);
+
+        examRepository.save(examModel);
```
Poderia retornar o exame agendado.

mathpablo committed:
Alterções feita conforme conversamos

mathpablo committed:
Separação do metodo scheduleExam

mathpablo committed:
Código corrigido
